Enhance/map finish control

braem opened the pull request:
Closes #277 

- Added convar that toggles if movement is possible after map finish (`mom_mapfinished_movement_enable`). This includes working teleports. These are done in [mom_player.cpp](https://github.com/momentum-mod/game/blob/develop/mp/src/game/server/momentum/mom_player.cpp)
- Added the ability to disable the mouse input on map finish panel after it is enabled. Acts as a toggle for mouse input now.
- Added a 15 second fadeout when `mom_mapfinished_movement_enable = 1`, as per the corresponding issue.

### Checklist
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

braem commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.h:
```diff
@@ -92,6 +95,9 @@ class CHudMapFinishedDialog : public CHudElement, public vgui::EditablePanel
 
     int m_iCurrentPage, m_iVelocityType;
+
+    float m_flTimeTillFadeout;
```
holdover from testing; unusued. Thought visual studio was suppose to catch these :(

braem commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -305,6 +338,7 @@ void CHudMapFinishedDialog::Reset()
     SetRunSaved(false);
     SetRunUploaded(false);
     m_bCanClose = false;
+    m_fPanelAlpha = 255.0f;
```
Switched to using Panel's alpha commands.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -296,6 +324,11 @@ void CHudMapFinishedDialog::OnMousePressed(MouseCode code)
+    {
+        // Can't disable mouse if specing
+        if (m_bNotSpec) SetMouseInputEnabled(false);
+    }
```
Was a little confused on whether `m_bIsGhost` meant you were watching a replay or spectating, which I now believe is both. Deleted `m_bNotSpec` to use `m_bIsGhost` instead.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -31,7 +35,9 @@ CHudMapFinishedDialog::CHudMapFinishedDialog(const char *pElementName) : CHudEle
     m_pRunStats = nullptr;
     m_bIsGhost = false;
     m_bCanClose = false;
+    m_bNotSpec = true;
```
Deleted this field and moved to `m_bIsGhost`

braem commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -47,6 +53,7 @@ CHudMapFinishedDialog::CHudMapFinishedDialog(const char *pElementName) : CHudEle
     m_pPlayReplayButton = new ImagePanel(this, "Replay_Icon");
     m_pRepeatButton = new ImagePanel(this, "Repeat_Button");
     m_pDetachMouseLabel = new Label(this, "Detach_Mouse", "#MOM_MF_DetachMouse");
+    m_pAttachMouseLabel = new Label(this, "Attach_Mouse", "#MOM_MF_AttachMouse");
```
Kept the text localizations but moved to just one label.

braem commented:
Made the requested changes.
Also made right clicks toggle whether the mouse is active in replays as well.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -20,6 +21,9 @@
 using namespace vgui;
 
+static MAKE_TOGGLE_CONVAR(mom_mapfinished_movement_enable, "0", FCVAR_CLIENTDLL | FCVAR_CLIENTCMD_CAN_EXECUTE | FCVAR_ARCHIVE,
+                          "Toggles being able to move. 0 = OFF, 1 = ON\n");
```
Can we get a better string for this? It's pretty ambiguous:
```suggestion
                          "Toggles being able to move after completing a run. 0 = OFF, 1 = ON\n");
```

Gocnak commented on mp/src/game/server/momentum/mom_player.cpp:
```diff
@@ -1573,14 +1573,13 @@ void CMomentumPlayer::TravelSpectateTargets(bool bReverse)
 {
     // slowdown when map is finished
-    if (m_Data.m_bMapFinished)
+    if (m_Data.m_bMapFinished && !ConVarRef("mom_mapfinished_movement_enable").GetBool())
```
I think it would be better to not call the `TweenSlowdownPlayer` function if the movement_enable var isn't enabled instead, saving you from creating and checking this ConVarRef every tick that the slowdown code would normally happen.

braem replied:
Fixed these calls, though not sure where to go to avoid calling this function. Looks like it is a callback function set up when the player spawns. I moved the ConVarRef into the constructor to avoid calls each tick.

braem commented:
Avoid those `ConVarRef` calls each tick in the `TweenSlowdownPlayer` and `Teleport` functions. Moved the call to the constructor instead.
Also added this to the `Teleport` function in `mom_replay_entity.cpp` in a similar way. I believe this corrects some unlikely edge case where the replay ghost hits a teleport after map finish.

Gocnak commented on mp/game/momentum/resource/momentum_english_ref.res:
```diff
@@ -218,7 +218,8 @@
         "MOM_MF_XPGainRank" "Rank XP Gained: %s1"
         "MOM_MF_PlayReplay" "Play Replay"
-        "MOM_MF_DetachMouse" "Right click to enable the mouse!"
+        "MOM_MF_MouseInactive" "Right click to enable the mouse!"
+        "MOM_MF_MouseActive" "Right click to disable the mouse!"
```
Now I'm starting to think this can just be one label:
```suggestion
        "MOM_MF_MouseToggle" "Right click to toggle mouse control!"
```

braem replied:
Good idea. Also don't need to worry about dynamically setting the label text now. Removed a useless override to `SetMouseInputEnabled` function thanks to that.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -336,10 +372,21 @@ void CHudMapFinishedDialog::SetVisible(bool bVisible)
         if (pSpecUI && pSpecUI->IsVisible() && ipanel()->IsMouseInputEnabled(pSpecUI->GetVPanel()))
             SetMouseInputEnabled(true);
+
+        if (mom_mapfinished_movement_enable.GetBool())
+        {
+            g_pClientMode->GetViewportAnimationController()->StopAnimationSequence(this, "FadeOutMapFinishedPanel");
+            g_pClientMode->GetViewportAnimationController()->StartAnimationSequence("FadeOutMapFinishedPanel");
+        }
+        else
+        {
+            g_pClientMode->GetViewportAnimationController()->StopAnimationSequence(this, "FadeOutMapFinishedPanel");
+        }
```
Both branches call stop animation for the `"FadeOutMapFinishedPanel"` animation, could be pulled out to above these two.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_mapfinished.cpp:
```diff
@@ -407,4 +454,14 @@ void CHudMapFinishedDialog::SetCurrentPage(int pageNum)
+
+    if (!mom_mapfinished_movement_enable.GetBool() || m_bIsGhost)
+    {
+        SetAlpha(255.0f);
```
`SetAlpha` takes an int:
```suggestion
        SetAlpha(255);
```

Do this for every instance.

braem replied:
Odd, the `SetAlpha` function does take an int, but the alpha value there, `m_flAlpha` is a float. It's implicit casting that int to a float x.x

braem replied:
Nevertheless I fixed this.

braem commented:
> 
> 
> What happens if players spectate (either from menu while in lobby or clicking the panel itself), does it still show? If so I don't think that should be intended.

It doesn't show in spectate no. Tested it with Exile to make sure. I was confusing spectating with watching a replay earlier.
